Remapping after thresholding

torms3 opened the pull request:
It is sometimes necessary to convert a real-valued detection map (e.g. glia) into an integer-valued semantic map, which can be used as input to semantic cuts in agglomeration. To this end, arbitrary remapping after thresholding has been added as postprocessing.

ranlu reviewed (commented):
Since you remap after thresholding, the remap table only need one value for true and one for false. In this case I prefer to add two options threshold operator, one for value above the threshold and one for value below the threshold. This way you do not need to use double eval to handle strange inputs, you can also check if the value given is valid for uint8, (an uninformed user might want to use -1 for false for example). You can also get rid of fastremap with simple numpy conditional indexing.

torms3 committed:
feat: add remapping after thresholding

It is sometimes necessary to convert a real-valued detection map (e.g. glia)
into an integer-valued semantic map, which can be used as input to
semantic cuts in agglomeration. To this end, arbitrary remapping after
thresholding has been added as postprocessing.

ranlu commented on chunkflow/chunk/base.py:
```diff
@@ -378,17 +378,26 @@ def crop_margin(self, margin_size: tuple = None, output_bbox: Bbox=None):
+            seg.array[seg.array == 1] = remap_high
+        if remap_low is not None:
+            assert remap_low >= 0 and remap_low < 256
+            seg.array[seg.array == 0] = remap_low
```
The code is problematic if someone decide to set remap_high = 0, then everything will be mapped to remap_low. It is safer to replace the condition `seg.array == 1/0` with `self >/<= threshold`.

torms3 replied:
That's a good catch. Will modify it.

ranlu commented on chunkflow/chunk/base.py:
```diff
@@ -378,17 +378,26 @@ def crop_margin(self, margin_size: tuple = None, output_bbox: Bbox=None):
+
+        # Optional remapping
+        if remap_high is not None:
+            assert remap_high >= 0 and remap_high < 256
```
You can be extra generic and use `numpy.iinfo` to figure out the the min and max if you want, but since you explicitly want uint8 already it does not matter.

ranlu commented on chunkflow/chunk/base.py:
```diff
@@ -378,17 +378,27 @@ def crop_margin(self, margin_size: tuple = None, output_bbox: Bbox=None):
-        seg = self > threshold
+    def threshold(self,
+                  threshold: float,
+                  remap_high: int = None,
```
I am not an expert of python type annotation, does the line actually pass the check? since you set `remap_high` to `None` which is not `int`?

torms3 replied:
I'm also new to Python type hinting, and you're right, [PEP484](https://peps.python.org/pep-0484/#union-types) says it's not recommended. It should be `remap_high: Optional[int] = None` instead. Will modify it.

Your meticulous review is truly useful! I'm really enjoying it.

ranlu commented on chunkflow/chunk/base.py:
```diff
@@ -378,17 +378,27 @@ def crop_margin(self, margin_size: tuple = None, output_bbox: Bbox=None):
+        # Optional remapping
+        iu8 = np.iinfo(np.uint8)
+        if remap_high is not None:
+            assert (remap_high >= iu8.min) and (remap_high <= iu8.max)
```
Is it possible to write it as `assert iu8.min <= remap_high <= iu8.max`?

torms3 replied:
Sure. I didn't know that's possible in Python.
